mne/beamformer/_compute_beamformer.py

```diff
@@ -144,52 +143,96 @@ def _normalized_weights(Wk, Gk, Cm_inv_sq, reduce_rank, nn, sk,
     .. [1] Sekihara & Nagarajan. Adaptive spatial filters for electromagnetic
            brain imaging (2008) Springer Science & Business Media
     """
-    norm_inv = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+    G_stacked_T = G.T.reshape(n_sources, n_orient, -1)
+    tmp = np.matmul(Cm_inv_sq, G)
+    tmp = tmp.reshape(-1, n_orient, n_sources, order='F').transpose(2, 0, 1)
+    norm_inv = np.matmul(G_stacked_T, tmp)
+
+    evals_denom, evecs_denom = np.linalg.eigh(norm_inv)
+    evals_denom = evals_denom[:, np.newaxis, :]
+    evals_sign = np.sign(evals_denom)
+    evals_denom *= evals_sign
+    evecs_denom *= evals_sign
     if reduce_rank:
         # Use pseudo inverse computation setting smallest
         # component to zero if the leadfield is not full rank
-        norm = _reg_pinv(norm_inv, rank=norm_inv.shape[0] - 1,
-                         svd_lwork=svd_lwork)[0]
+        # here we can use eigh to compute pinv; dmalt
+        # norm = _reg_pinv(norm_inv, rank=norm_inv.shape[0] - 1,
+        #                  svd_lwork=svd_lwork)[0]
+        U = np.zeros_like(evecs_denom)
+        U[:, :, 1:3] = (
+            evecs_denom[:, :, 1:3] / np.sqrt(evals_denom[:, :, 1:3])
+        )
     else:
         # Use straight inverse with full rank leadfield
         try:
-            norm = _repeated_inv(norm_inv, inv_lwork)
-        except LinAlgError:
+            # norm = _repeated_inv(norm_inv, inv_lwork)
+            with np.errstate(divide='raise'):
+                U = np.divide(evecs_denom, np.sqrt(evals_denom))
+        except (FloatingPointError, Exception):
             raise ValueError(
                 'Singular matrix detected when estimating spatial filters. '
                 'Consider reducing the rank of the forward operator by using '
                 'reduce_rank=True.'
             )
     # Reapply source covariance after inversion
-    norm *= sk
-    norm *= sk[:, np.newaxis]
-    power = np.dot(norm, np.dot(Wk, Gk))
+    # It seems like this does nothing since _prepare_beamformer_input
+    # is always called with the default loose=None; dmalt
```

Comment:
Although I don't understand, how this should work for beamformers were the constraints nontrivial. I'm not sure if the implemented loose orientation scheme is correct. For the MNE inverse it's quite clear since it fits nicely into the derivation as a prior sourse covariance but here I can't make my head around it.  Maybe I could read on it somewhere?

Reply:
> Maybe I could read on it somewhere?

There is no paper on it, it's why it's "hidden" currently :)

@britta-wstnr had a similar feeling that it might not be correct. Someone needs to test it. But having it in there makes it testable.

One thing that makes me think it might be correct is that its behavior at the limits is what you'd expect -- a `loose=0.999` looks just like `loose=1.` (free) and `loose=0.001` looks just like `loose=0.` (fixed), and it goes smoothly between these two extremes.

If you're interested in looking into it, there might be a paper in it for you!

Reply:
Ok, I see. If you've tested it and it gives reasonable results, I'll just leave it as is for now. Although another similar option would be to multiply power by the constraint, not just norm and it would be interesting to see how this can be inserted in the original optimization objective. Maybe I'll look into it at some point. Thanks!